### PR TITLE
[ENH] Categoricals improve

### LINKED ISSUE #927: [ENH] improve encode_categoricals
# Brief Description

- Simplify logic for `_encode_categoricals` in utils.py. A simple for-loop possibly
    - If user provides `categories` use as-is (order defaults to `appearance`)
    - `order` should be relevant only if `categories` is not provided.  
- Get rid of `As_categorical` helper function (tuple of `categories, order` is sufficient)

### PR DESCRIPTION
**Pull request by samukweku**
# PR Description

Please describe the changes proposed in the pull request:

-  Drop `As_Categorical` helper function
- Move checks into one function
- Avoid `copy` of the DataFrame at the start of the function ``df = df.copy()``

**This PR resolves #927.**

# PR Checklist

Please ensure that you have done the following:

1. [x] PR in from a fork off your branch. Do not PR from `<your_username>`:`dev`, but rather from `<your_username>`:`<feature-branch_name>`.

2. [x] If you're not on the contributors list, add yourself to `AUTHORS.rst`.

3. [ ] Add a line to `CHANGELOG.md` under the latest version header (i.e. the one that is "on deck") describing the contribution.
    - Do use some discretion here; if there are multiple PRs that are related, keep them in a single line.

# Automatic checks

There will be automatic checks run on the PR. These include:

- Building a preview of the docs on Netlify
- Automatically linting the code
- Making sure the code is documented
- Making sure that all tests are passed
- Making sure that code coverage doesn't go down.

# Relevant Reviewers

Please tag maintainers to review.

- @ericmjl

### CODE REVIEW

File: janitor/functions.py
Comment (ericmjl) [inline janitor/functions.py]: I think this would be misleading for readers of the code. The alternative values should be placed as comments after. Here's an example:

```suggestion
                    col2 = (categories = [values],
                            order="sort"  # or "appearance" or None
                           )
```

File: janitor/functions.py
Comment (ericmjl) [inline janitor/functions.py]: Same as above.

File: janitor/utils.py
Comment (ericmjl) [inline janitor/utils.py]: SEMBR please!

File: tests/functions/test_encode_categorical.py
Comment (ericmjl) [inline tests/functions/test_encode_categorical.py:24]: Thank you for doing this! The `JanitorError` was getting overused and becoming uninformative. Getting rid of `JanitorError` is a great idea.

**Review by ericmjl — CHANGES_REQUESTED** [review]
It's a big PR and I was reviewing it at 11 pm at night. My review is necessarily not going to be complete.

I think we should get another pair of eyes on this PR before merging.

File: janitor/functions.py
Comment (loganthomas) [inline janitor/functions.py]:
```suggestion
                        order="sort"  # or "appearance" or None
```

File: janitor/functions.py
Comment (loganthomas) [inline janitor/functions.py]: Does it matter that one of these is `tuple` and another is `Tuple`?

File: janitor/utils.py
Comment (loganthomas) [inline janitor/utils.py]: @ericmjl not sure if we are enforcing CapWord case for classes https://www.python.org/dev/peps/pep-0008/#class-names

File: janitor/utils.py
Comment (loganthomas) [inline janitor/utils.py:1897]: `nit`: if this is just finding uniques, could it be `pd.Index(set(df['column_name']))` or `pd.Index(df['column_name'].unique())`

Comment (samukweku) [reply janitor/utils.py:1897]: If the column has nulls, pd.unique wont drop them; factorize however will return non null values, which allows for proper testing of emptiness.

File: janitor/utils.py
Comment (samukweku) [reply janitor/utils.py]: first time for me. Would love to know what the accepted form is and I'll gladly modify.

File: janitor/utils.py
Comment (ericmjl) [reply janitor/utils.py:1897]: I see! That makes a lot of sense. Thanks @samukweku for educating!

File: janitor/utils.py
Comment (ericmjl) [reply janitor/utils.py]: Let's stick with Python convention. Thanks for picking it out, @loganthomas. @samukweku would you be kind enough to handle this piece? After that, be sure to send a ping either here or in our team chat so I can merge.

**Comment by samukweku** [comment]
@ericmjl @loganthomas made the edits  .... ready for your review

**Review by ericmjl — APPROVED** [review]

**Comment by ericmjl** [comment]
It's been reviewed by two people, so I'm going to merge!

**Pull request merged**
